core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java

```diff
@@ -4768,6 +4768,112 @@ public String getDefaultScheme() {
+              .setDescription("Failed to resolve name: " + resolutionError)
+              .setSeverity(ChannelTrace.Event.Severity.CT_WARNING)
+              .setTimestampNanos(timer.getTicker().read())
+              .build());
+
+      // no new trace events due to lastResolutionState already set to ERROR
+      assertThat(getStats(channel).channelTrace.events).hasSize(prevSize);
+    } finally {
+      LoadBalancerRegistry.getDefaultRegistry().deregister(mockLoadBalancerProvider);
+    }
+  }
+
+  @Test
+  public void testTraceEventsOnResultsCallFollowedByOnErrorCallWithAppliedServiceConfig()
+          throws Exception {
+    LoadBalancerRegistry.getDefaultRegistry().register(mockLoadBalancerProvider);
+    timer.forwardNanos(1234);
+    channelBuilder.maxTraceEvents(10);
+    try {
+      List<EquivalentAddressGroup> servers = new ArrayList<>();
+      servers.add(new EquivalentAddressGroup(socketAddress));
+      FakeNameResolverFactory nameResolverFactory =
+          new FakeNameResolverFactory.Builder(expectedUri).setServers(servers).build();
+      channelBuilder.nameResolverFactory(nameResolverFactory);
+      Map<String, Object> defaultServiceConfig =
+          parseConfig("{\"methodConfig\":[{"
+              + "\"name\":[{\"service\":\"SimpleService1\"}],"
+              + "\"waitForReady\":true}]}");
+      channelBuilder.defaultServiceConfig(defaultServiceConfig);
+      ManagedChannelServiceConfig managedChannelServiceConfig =
+          createManagedChannelServiceConfig(defaultServiceConfig, null);
+      nameResolverFactory.nextConfigOrError.set(
+          ConfigOrError.fromConfig(managedChannelServiceConfig));
+      createChannel();
+      int prevSize = getStats(channel).channelTrace.events.size();
+      Status resolutionError = Status.UNAVAILABLE
+          .withDescription("Initial Name Resolution error, using default service config");
+
+      nameResolverFactory.resolvers.get(0).listener.onError(resolutionError);
+
+      assertThat(getStats(channel).channelTrace.events).hasSize(prevSize + 1);
+      prevSize = getStats(channel).channelTrace.events.size();
+      ResolutionResult resolutionResult = ResolutionResult.newBuilder()
+          .setAddresses(Collections.singletonList(
+              new EquivalentAddressGroup(Arrays.asList(new SocketAddress() {
+              }, new SocketAddress() {
+              })))).build();
+      nameResolverFactory.resolvers.get(0).listener.onResult(resolutionResult);
+
+      assertEquals(channel.getConfigSelector().getClass().getName(),
+          managedChannelServiceConfig.getDefaultConfigSelector().getClass().getName());
+      assertEquals(channel.getLastServiceConfig().toString(),
+          managedChannelServiceConfig.toString());
+
+      // initial service config is already applied so it's not reapplied using the default service
+      // config here.
+      assertThat(getStats(channel).channelTrace.events).hasSize(prevSize + 2);
```

Comment:
How does this assertion for prevSize + 2 fit in with the above comment? The comment seems to imply that the count should not have increased.

Reply:
we need to delete this comment as it was added as pert of previous implementation. I will delete it

Reply:
Fixed